### PR TITLE
Wakeword release 0.1.2

### PR DESCRIPTION
**Pull request by hpsaturn**
Wakeword voice service for MALOS. The last version support:

    setup wakeword
    define lenguaje models path (pre-generated)
    define lenguaje dictionary path (pre-generated)
    define mic number (0-8)
    enable/disable pocketsphinx verbose debugging
    send and override configuration (hot-plug)
    disable voice recognition service (stop pshinx main thread)

Please see `README` file for installation instructions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,214 @@
...
+``` 
+echo "deb http://unstable-packages.matrix.one/ stable main" | sudo tee -a /etc/apt/sources.list
+sudo apt-get update
+sudo apt-get install matrix-creator-malos-wakeword --no-install-recommends
+```
+
+Nodejs and npm on RaspberryPi:
+
+``` 
+curl -sL https://deb.nodesource.com/setup_6.x | sudo bash -
```
Comment (arhuaco) [inline README.md:41]: Isn't 6.x too old? We're currently using 7 for other things.

File: README.md [hunk 2]
```diff
@@ -1 +1,214 @@
...
+sudo apt-get install nodejs
+```
+
+## Run sample DEMO
+
+Before, copy on your RaspberryPi language models and dictionary for run the DEMO or generated
```
Comment (arhuaco) [inline README.md]: First, copy...

File: README.md [hunk 3]
```diff
@@ -1 +1,214 @@
...
+```
+
+#### Config and start wakeupword service
+
+``` javascript
+function startWakeUpRecognition(){
+  console.log('<== config wakeword recognition..')
+  var wakeword_config = new matrixMalosBuilder.WakeWordParams;
+  wakeword_config.set_wake_word("MIA");
+  wakeword_config.set_lm_path("/home/pi/assets/9854.lm");
```
Comment (arhuaco) [inline README.md:136]: What instructions leave this file there? Same for the next file.

File: README.md [hunk 4]
```diff
@@ -1 +1,214 @@
...
+        setEverloop(0, 255, 100, 0, 0.05) 
+        break;
+      case "MIA RING ORANGE":
+        setEverloop(255, 77, 0, 0, 0.05) 
+        break;
+      case "MIA RING CLEAR":
+        setEverloop(0, 0, 0, 0, 0) 
+        break;
+      default:
+        text = "==> Not handled voice command";
```
Comment (arhuaco) [inline README.md:182]: Is text doing anything? If not just delete it... Or make it a console.log('voice command not known: ' + wakeWordData.wake_word).

File: src/driver_wakeword.cpp [hunk 5]
```diff
@@ -14,106 +14,124 @@
...
-#include "pocketsphinx.h"
 
 #include "./driver_wakeword.h"
 #include "./src/driver.pb.h"
 
-namespace {
-
-static ps_decoder_t *ps_;
-static cmd_ln_t *config_;
+namespace {}  // namespace
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: delete...

File: src/driver_wakeword.cpp [hunk 6]
```diff
@@ -14,106 +14,124 @@
...
-  tmo.tv_usec = ms * 1000;
+namespace matrix_malos {
 
-  select(0, NULL, NULL, NULL, &tmo);
-#endif
+bool WakeWordDriver::ProcessConfig(const DriverConfig &config) {
+  stopPipe();
+  WakeWordParams wakeword_params(config.wakeword());
+  if (wakeword_params.stop_recognition()) {
+    std::cout << "==> disable wakeword service.." << std::endl;
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: cerr

File: src/driver_wakeword.cpp [hunk 7]
```diff
@@ -14,106 +14,124 @@
...
+    return true;
+  }
+  loadParameters(wakeword_params);
+  if (validatePaths()) {
+    verbose = wakeword_params.enable_verbose();
+    enable = startPipe();
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: shouldn't this variable be named "enabled?". Not a big deal.

File: src/driver_wakeword.cpp [hunk 8]
```diff
@@ -14,106 +14,124 @@
...
-    }
-    sleep_msec(10);
-  }
-  ad_close(ad);
+void WakeWordDriver::loadParameters(WakeWordParams wakeword_params) {
+  channel = static_cast<int16_t>(wakeword_params.channel());
+  wakeword = std::string("" + wakeword_params.wake_word());
+  lm_path = std::string("" + wakeword_params.lm_path());
+  dic_path = std::string("" + wakeword_params.dic_path());
+  std::cout << "==> wakeword: " << wakeword << std::endl;
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: cerr

File: src/driver_wakeword.cpp [hunk 9]
```diff
@@ -14,106 +14,124 @@
...
+      "\" -inmic yes -adcdev mic_channel" + std::to_string(channel));
 
-namespace matrix_malos {
+  if (!verbose)
+    cmd = cmd + " 2> /dev/null";
 
-bool WakeWordDriver::ProcessConfig(const DriverConfig& config) {
-  // TODO @hpsaturn: Validate all the data that comes from the protos
-  WakeWordParams wakeword_params(config.wakeword());
+  std::cout << "Starting PocketSphinx thread.." << std::endl;
```
Comment (arhuaco) [inline src/driver_wakeword.cpp:76]: cerr

File: src/driver_wakeword.cpp [hunk 10]
```diff
@@ -14,106 +14,124 @@
...
+  returnMatch("voice recognition ready");
 
   return true;
 }
 
-bool WakeWordDriver::SendUpdate() {
+bool WakeWordDriver::stopPipe() {
+  if (sphinx_pipe_ != NULL && enable) {
+    enable = false;
+    std::cout << "Stoping PocketSphinx thread.." << std::endl;
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: cerr

File: src/driver_wakeword.cpp [hunk 11]
```diff
@@ -14,106 +14,124 @@
...
+    if (system(std::string("killall malos_psphinx").c_str()) == -1) {
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: system("killall malos_psphinx")

File: src/driver_wakeword.cpp [hunk 12]
```diff
@@ -14,106 +14,124 @@
...
+      return false;
+    }
+    sleep(2);
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: You should add a comment above the sleep call. Saying what the intention of the sleep is.

File: src/driver_wakeword.cpp [hunk 13]
```diff
@@ -14,106 +14,124 @@
...
+  while (fgets(buff, sizeof(buff), sphinx_pipe_) != NULL && enable) {
+    std::string sbuff = buff;
+    std::size_t found = sbuff.find(match);
+    if (found != std::string::npos) {
+      std::cout << "[SEND] recognition " << sbuff;
+      sbuff = sbuff.substr(sbuff.find_first_of(" \t") + 1);
+      sbuff = sbuff.erase(sbuff.length() - 1);
+      returnMatch(sbuff);
+    } else {
+      std::cout << "[SKIP] recognition " << sbuff;
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: stderr

File: src/driver_wakeword.cpp [hunk 14]
```diff
@@ -14,106 +14,124 @@
...
+void WakeWordDriver::PocketSphinxProcess() {
+  char buff[512];
+  char match[100];
+  snprintf(match, sizeof(match), "match: %s", wakeword.c_str());
+
+  // processing pipe output
+  while (fgets(buff, sizeof(buff), sphinx_pipe_) != NULL && enable) {
+    std::string sbuff = buff;
+    std::size_t found = sbuff.find(match);
+    if (found != std::string::npos) {
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: if (sbuff.find(match) != std::string::npos) {
      // A match was found

File: src/driver_wakeword.cpp [hunk 15]
```diff
@@ -14,106 +14,124 @@
...
+      std::cout << "[SEND] recognition " << sbuff;
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: stderr

File: src/driver_wakeword.cpp [hunk 16]
```diff
@@ -14,106 +14,124 @@
...
+
+bool WakeWordDriver::validatePaths() {
+  if (!exists_path(dic_path))
+    return false;
+  if (!exists_path(lm_path))
+    return false;
   return true;
 }
+
+inline bool WakeWordDriver::exists_path(const std::string &name) {
```
Comment (arhuaco) [inline src/driver_wakeword.cpp]: You could define this function in the anonymous namespace, above. No need to make it a class member.

File: src/driver_wakeword.h [hunk 17]
```diff
@@ -18,37 +18,56 @@
...
   }
 
   // Read configuration from the outside workd.
-  bool ProcessConfig(const DriverConfig& config) override;
+  bool ProcessConfig(const DriverConfig &config) override;
+
+private:
+  // Thread that read events from pocketsphinx
+  void PocketSphinxProcess();
+
```
Comment (arhuaco) [inline src/driver_wakeword.h]: You should add a small comment before each method and member.

File: src/malos_psphinx.cpp [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+ * NOR ITS EMPLOYEES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ *
+ * continuous.c - Simple pocketsphinx command-line application to test
```
Comment (arhuaco) [inline src/malos_psphinx.cpp:35]: Mot making many comments to this file as it is coming from somewhere else.

File: README.md (hunk 1 above)
Comment (hpsaturn) [reply README.md:41]: recomendation from @maciekrb

File: README.md (hunk 3 above)
Comment (hpsaturn) [reply README.md:136]: is a language models and voice commands dictionary, pre-generated with `lmtool` from PocketSphinx webpage [1], the idea from next release is remove this dependence 

[lmtool](http://www.speech.cs.cmu.edu/tools/lmtool-new.html)

File: README.md (hunk 4 above)
Comment (hpsaturn) [reply README.md:182]: yea is a bug! done.

File: src/malos_psphinx.cpp (hunk 18 above)
Comment (hpsaturn) [reply src/malos_psphinx.cpp:35]: but this file is a product from old PR that you already reviewed

Comment (hpsaturn) [reply src/malos_psphinx.cpp:35]: (ready all comments, thanks)

**Pull request merged**